programs/survival/silk_blockstates.sc

```diff
@@ -0,0 +1,162 @@
+///
+// Silky Blockstates
+// by BisUmTo
+// (Carpet Mod 1.4.11)
+//
+// While sneaking, the silk_touch enchantment saves the Blockstates and the Blockdata of mined blocks.
+// It doesn't apply to containers and blacklisted blocks.
+///
+
+__config() -> {'stay_loaded' -> true, 'scope' -> 'global'};
+
+global_drop_in_creative = true;
+global_create_item_whitelist = ['^spawner$', '^cake$'];
+global_preserve_block_state_blacklist = ['_bed$', '_door$', '^sticky_piston$', '^piston$', '^bee_nest$', '_banner$', '^beehive$', '^redstone_wire$'];
+global_preserve_block_data_blacklist = ['^bee_nest$', '^beehive$', '^campfire$', '^soul_campfire$', '^lectern$', '^jukebox$', '_banner$', '^player_head$', '_bed$'];
+global_need_forced_placement = ['^spawner$','_sign$'];
+
+// returns a map with all block_properties of the block and the relative value
+block_state(block) -> (
+    keys = block_properties(block) ;
+    values = map(keys, property(block, _));
+    m = {};
+    for(keys, m:_ = values:_i);
+    return(m);
+);
+
+// returns the escaped version of the string (\->\\, d->\d)
+escape_string(string,d) -> (
+    replace(replace(string, '\\\\', '\\\\\\\\'), d, '\\\\'+d)
+);
+
+// returns true if the block is repleaceable while trying to place a block in its position
+replaceable(block) ->(
+    ['air', 'bubble_column', 'cave_air', 'crimson_roots', 'dead_bush', 'fern', 'fire', 'grass', 'large_fern', 'lava','nether_sprout','sea_grass', 'soul_fire', 'structure_void', 'tall_grass', 'vines', 'void_air', 'warped_roots', 'water'] ~ block != null || (block == 'snow' && property(block, 'layers') == 1)
```

Comment:
suggest defining these blocks outside and as a set, not list so `['foo', 'bar'] ~ block` => `global_replaceable = {'foo', 'bar'}; has(global_replaceable, str(block))`

Probably in hindsight, won't make a big difference but hey.